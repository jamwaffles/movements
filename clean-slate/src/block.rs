//! Parse a block (line) populated with [`Word`]s.

use crate::spanned_word::{Span, Spanned, Word};
use nom::character::complete::space0;
use nom::{multi::many0, sequence::preceded, IResult};

#[derive(Debug)]
pub struct Block<'a> {
    words: Vec<Spanned<'a, Word>>,
}

// TODO: Trait?
impl<'a> Block<'a> {
    pub fn parse(i: Span<'a>) -> IResult<Span, Self> {
        let (i, words) = many0(preceded(space0, Word::parse))(i)?;

        Ok((i, Self { words }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn newline() {
        insta::assert_debug_snapshot!(Block::parse("G0 G4 P2.5 ; line comment".into()));
    }
}
