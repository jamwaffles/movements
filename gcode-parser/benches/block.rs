use criterion::{black_box, criterion_group, criterion_main, Criterion};
use gcode_parser::{block::block, ParseInput};

fn block_simple(c: &mut Criterion) {
    c.bench_function("block simple", |b| {
        b.iter(|| block(black_box(ParseInput::new("G1 X10.3 Y20.1 Z30.2"))))
    });
}

fn block_delete(c: &mut Criterion) {
    c.bench_function("block delete", |b| {
        b.iter(|| block(black_box(ParseInput::new("/ G1 X10.3 Y20.1 Z30.2"))))
    });
}

fn block_line_number(c: &mut Criterion) {
    c.bench_function("block line number", |b| {
        b.iter(|| block(black_box(ParseInput::new("N1234 G1 X10.3 Y20.1 Z30.2"))))
    });
}

fn block_line_number_and_delete(c: &mut Criterion) {
    c.bench_function("block line number and delete", |b| {
        b.iter(|| block(black_box(ParseInput::new("/ N1234 G1 X10.3 Y20.1 Z30.2"))))
    });
}

criterion_group!(
    benches,
    block_simple,
    block_delete,
    block_line_number,
    block_line_number_and_delete
);
criterion_main!(benches);
