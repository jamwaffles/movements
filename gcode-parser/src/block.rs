use crate::token::{token, Token, TokenType};
use crate::ParseInput;
use nom::character::complete::space0;
use nom::multi::many0;
use nom::sequence::terminated;
use nom::IResult;

#[derive(Debug, Clone, PartialEq)]
pub struct Block {
    /// Whether the line has a leading `/` block delete character present
    pub block_delete: bool,

    /// All tokens in this block, including optional block delete and line numbers
    pub tokens: Vec<Token>,
}

/// Parse a block (single line of gcode)
pub fn block(i: ParseInput) -> IResult<ParseInput, Block> {
    // TODO: Fix and benchmark `separated_list()` if I can get it to support zero length separators
    let (i, tokens) = many0(terminated(token, space0))(i)?;

    Ok((
        i,
        Block {
            block_delete: tokens.first().map(|t| &t.token) == Some(&TokenType::BlockDelete),
            tokens,
        },
    ))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{coord::Coord, motion::Motion};
    use crate::{rem, tok};

    #[test]
    fn block_delete() {
        assert_eq!(
            block(ParseInput::new("/ G1 X10")),
            Ok((
                rem!("", 8),
                Block {
                    block_delete: true,
                    tokens: vec![
                        tok!(TokenType::BlockDelete, 0, 1),
                        tok!(TokenType::Motion(Motion::Feed), 2, 4),
                        tok!(TokenType::Coord(Coord::with_x(10.0)), 5, 8)
                    ]
                }
            ))
        );
    }

    #[test]
    fn line_number() {
        assert_eq!(
            block(ParseInput::new("N1234 G1 X10")),
            Ok((
                rem!("", 12),
                Block {
                    block_delete: false,
                    tokens: vec![
                        tok!(TokenType::LineNumber(1234u32), 0, 5),
                        tok!(TokenType::Motion(Motion::Feed), 6, 8),
                        tok!(TokenType::Coord(Coord::with_x(10.0)), 9, 12)
                    ]
                }
            ))
        );
    }

    #[test]
    fn line_number_and_block_delete() {
        assert_eq!(
            block(ParseInput::new("/ N1234 G1 X10")),
            Ok((
                rem!("", 14),
                Block {
                    block_delete: true,
                    tokens: vec![
                        tok!(TokenType::BlockDelete, 0, 1),
                        tok!(TokenType::LineNumber(1234u32), 2, 7),
                        tok!(TokenType::Motion(Motion::Feed), 8, 10),
                        tok!(TokenType::Coord(Coord::with_x(10.0)), 11, 14),
                    ]
                }
            ))
        );
    }
}
