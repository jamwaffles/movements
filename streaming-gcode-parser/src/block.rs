use crate::statement::Statement;
use crate::Span;
use nom::{
    branch::alt,
    bytes::streaming::tag,
    bytes::streaming::tag_no_case,
    bytes::streaming::take_till,
    bytes::streaming::take_until,
    bytes::streaming::take_while,
    character::streaming::char,
    character::streaming::digit1,
    character::streaming::not_line_ending,
    character::streaming::one_of,
    character::streaming::space0,
    character::streaming::space1,
    character::{
        complete::line_ending,
        streaming::{alpha1, anychar, multispace0},
    },
    combinator::map,
    combinator::map_res,
    combinator::opt,
    combinator::peek,
    combinator::recognize,
    combinator::{cond, verify},
    combinator::{eof, map_opt},
    error::ParseError,
    multi::many0,
    multi::many1,
    multi::many_m_n,
    multi::separated_list0,
    sequence::delimited,
    sequence::preceded,
    sequence::{separated_pair, terminated},
    IResult, InputIter, InputLength, Parser,
};
use std::str::FromStr;

pub fn end_of_line(i: Span) -> IResult<Span, Span> {
    if i.is_empty() {
        Ok((i, i))
    } else {
        line_ending(i)
    }
}

#[derive(Debug, PartialEq, Clone, Default)]
pub struct Block {
    block_delete: bool,
    line_number: Option<u32>,
    // TODO: Un-pub
    pub words: Vec<Statement>,
}

fn parse_words(i: Span) -> IResult<Span, Vec<Statement>> {
    let mut i = i;
    let mut res = Vec::new();

    loop {
        match preceded(space0, Statement::parse)(i) {
            Err(nom::Err::Error(_)) => {
                break Ok((i, res));
            }
            Err(e) => {
                if res.is_empty() {
                    break Err(e);
                } else {
                    break Ok((i, res));
                }
            }
            Ok((i1, o)) => {
                res.push(o);
                i = i1;
            }
        }
    }
}

impl Block {
    pub fn words(words: Vec<Statement>) -> Self {
        Self {
            block_delete: false,
            line_number: None,
            words,
        }
    }

    fn parse_block_delete(i: Span) -> IResult<Span, Option<()>> {
        opt(map(char('/'), |_| ()))(i)
    }

    fn parse_line_number(i: Span) -> IResult<Span, Option<u32>> {
        opt(map_res(
            separated_pair(tag_no_case("N"), space0, digit1),
            |(_, number): (_, Span)| u32::from_str(&number.to_string()),
        ))(i)
    }

    pub fn parse(i: Span) -> IResult<Span, Self> {
        let (i, block_delete) = Self::parse_block_delete(i)?;

        let (i, line_number) = Self::parse_line_number(i)?;

        let (i, words) = parse_words(i)?;

        Ok((
            i,
            Self {
                block_delete: block_delete.is_some(),
                line_number,
                words,
            },
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        assert_parse,
        modal_groups::{DistanceMode, Units},
    };

    #[test]
    fn empty_line() {
        assert_parse!(Block::parse, "\n", ("\n", Block::default()));
    }

    #[test]
    fn check_block() {
        assert_parse!(
            Block::parse,
            "g21 g90;",
            (
                ";",
                Block::words(vec![
                    Statement::Units(Units::Mm),
                    Statement::DistanceMode(DistanceMode::Absolute),
                ])
            )
        );
    }

    #[test]
    fn check_block_no_spaces() {
        assert_parse!(
            Block::parse,
            "G21G90;",
            (
                ";",
                Block::words(vec![
                    Statement::Units(Units::Mm),
                    Statement::DistanceMode(DistanceMode::Absolute),
                ])
            )
        );
    }
}
