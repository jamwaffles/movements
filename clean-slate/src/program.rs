//! Parse a block (line) populated with [`Word`]s.

use crate::block::Block;
use crate::Span;
use nom::character::complete::line_ending;
use nom::error::{ContextError, ParseError};
use nom::multi::separated_list0;
use nom::IResult;

#[derive(Debug)]
pub struct Program<'a> {
    blocks: Vec<Block<'a>>,
}

// TODO: Trait?
impl<'a> Program<'a> {
    pub fn parse<E>(i: Span<'a>) -> IResult<Span, Self, E>
    where
        E: ParseError<Span<'a>> + ContextError<Span<'a>>,
    {
        let (i, blocks) = separated_list0(line_ending, Block::parse::<E>)(i)?;

        Ok((i, Self { blocks }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use nom::error::ErrorKind;

    #[test]
    fn newline() {
        let program = r#"(begin)
        G0
        G4 P2.5
        G0
        ;end"#;

        insta::assert_debug_snapshot!(Program::parse::<(_, ErrorKind)>(program.into()));
    }
}
