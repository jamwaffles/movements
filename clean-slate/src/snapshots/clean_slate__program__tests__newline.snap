---
source: clean-slate/src/program.rs
expression: "Program::parse(program.into())"

---
Ok(
    (
        LocatedSpan {
            offset: 58,
            line: 5,
            fragment: "",
            extra: (),
        },
        Program {
            blocks: [
                Block {
                    words: [
                        Spanned {
                            start: LocatedSpan {
                                offset: 0,
                                line: 1,
                                fragment: "",
                                extra: (),
                            },
                            end: LocatedSpan {
                                offset: 7,
                                line: 1,
                                fragment: "",
                                extra: (),
                            },
                            item: Comment(
                                Comment {
                                    kind: Inline,
                                    comment: "begin",
                                },
                            ),
                        },
                    ],
                },
                Block {
                    words: [
                        Spanned {
                            start: LocatedSpan {
                                offset: 16,
                                line: 2,
                                fragment: "",
                                extra: (),
                            },
                            end: LocatedSpan {
                                offset: 18,
                                line: 2,
                                fragment: "",
                                extra: (),
                            },
                            item: Motion(
                                Rapid,
                            ),
                        },
                    ],
                },
                Block {
                    words: [
                        Spanned {
                            start: LocatedSpan {
                                offset: 27,
                                line: 3,
                                fragment: "",
                                extra: (),
                            },
                            end: LocatedSpan {
                                offset: 34,
                                line: 3,
                                fragment: "",
                                extra: (),
                            },
                            item: NonModal(
                                Dwell {
                                    duration: 2.5s,
                                },
                            ),
                        },
                    ],
                },
                Block {
                    words: [
                        Spanned {
                            start: LocatedSpan {
                                offset: 43,
                                line: 4,
                                fragment: "",
                                extra: (),
                            },
                            end: LocatedSpan {
                                offset: 45,
                                line: 4,
                                fragment: "",
                                extra: (),
                            },
                            item: Motion(
                                Rapid,
                            ),
                        },
                    ],
                },
                Block {
                    words: [
                        Spanned {
                            start: LocatedSpan {
                                offset: 54,
                                line: 5,
                                fragment: "",
                                extra: (),
                            },
                            end: LocatedSpan {
                                offset: 58,
                                line: 5,
                                fragment: "",
                                extra: (),
                            },
                            item: Comment(
                                Comment {
                                    kind: Block,
                                    comment: "end",
                                },
                            ),
                        },
                    ],
                },
            ],
        },
    ),
)
